ext/srt/gstsrtobject.c

```diff
@@ -829,6 +850,84 @@ thread_func (gpointer data)
   }
 }
 
+static void
+parse_stream_id (const gchar * stream_id, gchar ** username, gchar ** resource)
+{
+  const gchar STREAM_ID_PREFIX[] = "#!::";
+  gchar **keys;
+  gchar **it;
+
+  if (!g_str_has_prefix (stream_id, STREAM_ID_PREFIX)) {
+    return;
+  }
+  if (username) {
+    *username = NULL;
+  }
+  if (resource) {
+    *resource = NULL;
+  }
+  keys = g_strsplit (stream_id + sizeof (STREAM_ID_PREFIX) - 1, ",", -1);
+  for (it = keys; *it; ++it) {
+    gchar **keyval;
+    keyval = g_strsplit (*it, "=", 2);
+    if (keyval && keyval[0] && keyval[1]) {
+      if (g_str_equal (keyval[0], "u") && username) {
+        g_clear_pointer (username, g_free);
+        *username = g_strdup (keyval[1]);
+      } else if (g_str_equal (keyval[0], "r") && resource) {
+        g_clear_pointer (resource, g_free);
+        *resource = g_strdup (keyval[1]);
+      }
+    }
+    g_strfreev (keyval);
+  }
+  g_strfreev (keys);
+}
+
+static GSocketAddress *
+peeraddr_to_g_socket_address (const struct sockaddr *peeraddr) {
+  gsize peeraddr_len;
+
+  switch (peeraddr->sa_family) {
+    case AF_INET:
+      peeraddr_len = sizeof (struct sockaddr_in);
+      break;
+    case AF_INET6:
+      peeraddr_len = sizeof (struct sockaddr_in6);
+      break;
+    default:
+      g_warning ("Unsupported address family %d", peeraddr->sa_family);
+      return NULL;
+  }
+  return g_socket_address_new_from_native ((gpointer) peeraddr, peeraddr_len);
+}
+
+static gint
+srt_listen_callback_func(GstSRTObject * self, SRTSOCKET sock, int hs_version,
+    const struct sockaddr* peeraddr, const char* stream_id) {
+  g_autoptr (GSocketAddress) addr = peeraddr_to_g_socket_address (peeraddr);
+  gchar *username = NULL;
+  g_autofree gchar *resource = NULL;
+
+  if (self->authentication) {
+    gboolean authenticated = FALSE;
+
+    parse_stream_id (stream_id, &username, &resource);
+    /* notifying caller-connecting */
+    g_signal_emit_by_name (self->element, "caller-connecting", addr,
```

Comment:
If this is to be upstreamed, I think we should be more generic than we are in `HwangsaeRelay` and don't expect the Stream ID conforms to [SRT Access Control Guidelines](https://github.com/Haivision/srt/blob/master/docs/AccessControl.md).

That means I wouldn't parse username and resource inside `srt_listen_callback_func` but passed `stream_id` into `"caller-connecting"` signal verbatim. Any string parsing and inspection should be done by the application.

Reply:
Removed username and resource parsing. Passed stream_id to "caller-connecting" callback.